app/utils/sparse-fieldsets.ts

```diff
@@ -0,0 +1,70 @@
+import { camelize, underscore } from '@ember/string';
+import ModelRegistry from 'ember-data/types/registries/model';
+import { pluralize, singularize } from 'ember-inflector';
+import { Links } from 'jsonapi-typescript';
+
+import { Resource } from 'osf-api';
+
+// Automatic type safety for attributes and relationships appears prohibitively expensive.
+// Instead, use `unknown` to force explicit types.
+export type SparseModel = {
+    id: string | number;
+    modelName: string;
+    links?: Links;
+} & Record<string, unknown>;
+
+export type SparseFieldset = {
+    [K in keyof ModelRegistry]?: string[]
+};
+
+function mapObject<T, R>(
```

Comment:
What is this doing exactly? Just mapping values to keys?

Reply:
It creates a new object with the keys and values each mapped by a different function. In pseudocode:
```
mapObject({ k: v }, f, g) => { f(k): g(v) }
```
I found myself repeating that pattern several times (and it was pretty hard to read), so I pulled it out into a local util.

Reply:
What happened to mapObject?

Reply:
Oh sorry, shouldn't have force-pushed. I renamed it to `mapKeysAndValues`, because that seemed like a more understandable name.

Reply:
Gotcha, nw